### PR TITLE
refactor: add `join_unwind` to `SpawnedTask`

### PR DESCRIPTION
**Pull request by DDtKey**
## Which issue does this PR close?

Simple refactoring after #9318 

In order to avoid duplication of explicit handling it seems logical to have such method.

I thought to add this logic to `join` but there are calls with additional logic (usually returns `DatafusionError::Execution` instead of unwind)

## Rationale for this change

I noticed the same pattern spread across the repo. 
It also easy to forget and hard to refactor.

So could be useful

## Are there any user-facing changes?
No

### CODE REVIEW

File: datafusion/physical-plan/src/common.rs [hunk 1]
```diff
@@ -204,12 +204,24 @@ impl<R: 'static> SpawnedTask<R> {
...
+    /// Joins the task, returning the result of join (`Result<R, JoinError>`).
     pub async fn join(mut self) -> Result<R, JoinError> {
         self.inner
             .join_next()
             .await
             .expect("`SpawnedTask` instance always contains exactly 1 task")
     }
+
+    /// Joins the task and unwinds the panic if it happens.
+    pub async fn join_unwind(self) -> R {
```
Comment (DDtKey) [inline datafusion/physical-plan/src/common.rs:216]: Alternatively, we can always unwind in `join`. But for specific cases have `raw_join` or something like that

File: datafusion/core/src/datasource/file_format/parquet.rs [hunk 2]
```diff
@@ -952,31 +932,21 @@ async fn concatenate_parallel_row_groups(
...
+        let (serialized_columns, cnt) = result?;
+        row_count += cnt;
+        for chunk in serialized_columns {
+            chunk.append_to_row_group(&mut rg_out)?;
+            let mut buff_to_flush = merged_buff.buffer.try_lock().unwrap();
+            if buff_to_flush.len() > BUFFER_FLUSH_BYTES {
+                object_store_writer
+                    .write_all(buff_to_flush.as_slice())
+                    .await?;
+                buff_to_flush.clear();
```
Comment (devinjdangelo) [inline datafusion/core/src/datasource/file_format/parquet.rs:946]: This block in particular is much easier to read and understand now :+1:

File: datafusion/physical-plan/src/common.rs (hunk 1 above)
Comment (devinjdangelo) [reply datafusion/physical-plan/src/common.rs:216]: I think the existing naming makes sense to me i.e. call `join_unwind` if you don't want to manually handle the join error.

**Review by devinjdangelo — APPROVED** [review]
Thank you for the refactor and cleaning up this code. It is much easier to read now!

I am wondering if ideally these tasks _should_  immediately call `resume_unwind` when a child task panics, or if they should instead return an appropriate `DataFusionError`via one of the macros that add stack trace information. Perhaps the latter would give us more descriptive and consistent error messages. Regardless, I think this PR is a step in a positive direction so would not need to wait on this even if we decided returning errors would be better.

**Comment by devinjdangelo** [comment]
Also be aware #9414 was recently merged and caused some conflicts since `SpawnedTask` was moved.

**Comment by DDtKey** [comment]
> Also be aware https://github.com/apache/arrow-datafusion/pull/9414 was recently merged and caused some conflicts since SpawnedTask was moved.

Thanks! Rebased from onto main branch

File: datafusion/common_runtime/src/common.rs [hunk 3]
```diff
@@ -51,10 +51,22 @@ impl<R: 'static> SpawnedTask<R> {
...
 
+    /// Joins the task, returning the result of join (`Result<R, JoinError>`).
     pub async fn join(mut self) -> Result<R, JoinError> {
         self.inner
             .join_next()
             .await
             .expect("`SpawnedTask` instance always contains exactly 1 task")
     }
+
+    /// Joins the task and unwinds the panic if it happens.
```
Comment (alamb) [inline datafusion/common_runtime/src/common.rs:62]: 👍

File: datafusion/common_runtime/src/common.rs [hunk 4]
```diff
@@ -51,10 +51,22 @@ impl<R: 'static> SpawnedTask<R> {
...
+    pub async fn join_unwind(self) -> R {
+        self.join().await.unwrap_or_else(|e| {
+            if e.is_panic() {
+                std::panic::resume_unwind(e.into_panic());
+            } else {
+                unreachable!();
```
Comment (alamb) [inline datafusion/common_runtime/src/common.rs:68]: whi is this unreachable? Is there no other way for a Task::join to error?

**Review by alamb — APPROVED** [review]
Thank you @DDtKey  for the cleanup and @devinjdangelo for the review

File: datafusion/common_runtime/src/common.rs (hunk 4 above)
Comment (DDtKey) [reply datafusion/common_runtime/src/common.rs:68]: Hm, that's the good question. I just kept the logic as it used to be.

But in fact `JoinError` might be caused either by panic or cancellation of task.

And I think it's reachable in general, but if it was cancelled - likely it's expected or was due to another propagated error. 

For now I can suggest to remove `else { unreachable!() }`, but in long term we can think about smarter error handling for such cases

Comment (DDtKey) [reply datafusion/common_runtime/src/common.rs]: I've updated this block, because actually it's really confusing. Added some notes to code

Comment (DDtKey) [reply datafusion/common_runtime/src/common.rs]: I've update added a comment to the code to describe the behavior.

But in general as @devinjdangelo mentioned, we could improve error handling in the future

**Comment by alamb** [comment]
Thanks again @DDtKey  and @devinjdangelo  for the review

**Pull request merged**